promql: add sort_by_label and sort_by_label_desc functions

galexrt opened the pull request:
This adds functions to sort a vector by its label value.

Based on https://github.com/prometheus/prometheus/pull/1533

Signed-off-by: Alexander Trost <galexrt@googlemail.com>

***

There was some discussion around adding this functionality to PromQL/ Prometheus in the original PR #1533, though based on the various comments from users and now myself stumbling upon this while trying to visualize certain metrics sorted by label value, I would gladly see this topic revisited.

roidelapluie commented:
How difficult would it be to make this a vararg function to be able to sort by multiple labels?

`sort_by_label(up, "job","instance")`

galexrt commented:
@roidelapluie I can look into making it into a vararg function to be able to sort by multiple labels. Is there an existing function that already accepts a list of strings as an argument that I can look at?

roidelapluie commented:
`label_join()`

galexrt commented:
@roidelapluie The `sort_by_label` and `sort_by_label_desc` functions now accept a list of labels and is sorting the vector accordingly (see `functions.tests`).

juliusv commented on docs/querying/functions.md:
```diff
@@ -401,6 +401,14 @@ in ascending order.
 
+## `sort_by_label()`
+
+`sort_by_label(v instant-vector, label string, ...)` returns vector elements sorted by their label values and sample value in case of label values being equal, in ascending order.
```
Hmm, I wonder if we shouldn't keep this completely independent from the sample value and instead use other label values to break the sorting tie? That way, sort results would be stable over time, given the same set of time series.

roidelapluie replied:
I agree with Julius. We could also decide not to sort at all in that case, so you can do sort_by_label(sort(up),"job")

juliusv replied:
Yeah, that sounds even better.

galexrt replied:
So I should remove the "fallback" to comparing the value in the `sort_by_label` functions? This might make the test unstable, though I have to re-check if that happens without this "fallback" compare.

roidelapluie replied:
The tests should be stable even without this

roidelapluie replied:
we should make it clear that labels are sorted lexicographically

galexrt replied:
@roidelapluie Without the value comparison the test fails no matter in what order the result in the test(s) are placed.

roidelapluie replied:
That is strange. The order should be predictible.

galexrt replied:
It seems that it is not predictable as every X run the test fails and complains about the order of the result.

How should we approach this that we can hopefully get this functionality integrated?

galexrt replied:
From my perspective, the fix here is to sort the values using the `sort`/`sort_desc` functions and then do the sort by label(s).

juliusv commented on promql/functions.go:
```diff
@@ -284,6 +284,24 @@ func funcSortDesc(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
 
+// === sort_by_label(vector parser.ValueTypeVector, label parser.ValueTypeString...) Vector ===
+func funcSortByLabel(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) Vector {
+	// Incase the labels are the same, NaN should sort to the bottom, so take
```
"Incase" -> "In case"

juliusv commented on promql/functions.go:
```diff
@@ -1218,6 +1238,78 @@ func (s *vectorByReverseValueHeap) Pop() interface{} {
+
+func (s vectorByLabelHeap) Less(i, j int) bool {
+	// Iterate over each given label
+	for k := 0; k < len(s.labels); k++ {
```
This would be more Go-y:

```suggestion
	for _, label := range s.labels {
```

roidelapluie commented on promql/functions.go:
```diff
@@ -1218,6 +1238,78 @@ func (s *vectorByReverseValueHeap) Pop() interface{} {
+			continue
+		case 1:
+			return false
+		}
```
we need a return at the end

bboreham commented:
Note that range queries are always sorted by label sets, right at the end:
https://github.com/prometheus/prometheus/blob/8b863c42dd956d35d18a7a0b39c89c86adf7cebf/promql/engine.go#L722

Perhaps you could mention in the documentation this is intended for instant queries.

juliusv commented:
> Note that range queries are always sorted by label sets, right at the end:
https://github.com/prometheus/prometheus/blob/8b863c42dd956d35d18a7a0b39c89c86adf7cebf/promql/engine.go#L722
>
> Perhaps you could mention in the documentation this is intended for instant queries.

That same comment would apply to `sort()` and `sort_desc()` as well. In general, sorting is a no-op for range queries, as output points are always sorted by the X and Y axis there (unless someone cares about a specific legend item sorting, which I haven't heard yet).

I think we can either omit that comment on the new functions as well, or add a comment to the existing sort functions too. Something like:

> Please note that sorting functions only affect the results of instant queries, as range query results always have a fixed output ordering.

bboreham commented:
I think it would be easy to work out that `sort` (by value) is not a good idea over a range, where different series can be the highest value at different points in time. 

Whereas `sort_by_label` is something you might reasonably want for a range query, e.g. to change the order in which lines are arranged in a dashboard. Hence it should be documented that it won't do what you asked.

juliusv commented:
Ok yeah, that makes sense. Or we could actually make it work for range queries as well, but I doubt that's worth it. So then let's just add a comment for the new functions.

bboreham commented on promql/functions.go:
```diff
@@ -1218,6 +1238,78 @@ func (s *vectorByReverseValueHeap) Pop() interface{} {
 	return el
 }
 
+type vectorByLabelHeap struct {
```
What does the word "Heap" denote here?

bboreham commented on promql/functions.go:
```diff
@@ -284,6 +284,24 @@ func funcSortDesc(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
+	// Incase the labels are the same, NaN should sort to the bottom, so take
+	// ascending sort with NaN first and reverse it.
+	byLabelSorter := vectorByLabelHeap{vector: vals[0].(Vector), labels: stringSliceFromArgs(args[1:])}
+	sort.Sort(sort.Reverse(byLabelSorter))
```
`slices.SortFunc` is probably neater now.

bboreham reviewed (commented):
Hi, we discussed this PR at the [Prometheus dev summit](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?usp=sharing) today; we would like to move forward, but would like to ask you to put these functions behind a feature flag.

I had a couple of comments about the implementation too.

roidelapluie commented:
cc @galexrt are you still willing to work on this, rebase this pull request and put the functions behind a feature flag? Thanks!

galexrt commented:
> cc @galexrt are you still willing to work on this, rebase this pull request and put the functions behind a feature flag? Thanks!

Oh that totally fell of my radar, yes I am willing to work on that. Do you have a link to a feature flag example somewhere that I can use as an example?

roidelapluie commented:
You can look at the engine ops EnableNegativeOffset which used to be a feature flag

galexrt commented:
> You can look at the engine ops EnableNegativeOffset which used to be a feature flag

So should I dynamically add the `sort_by_label` and `sort_by_label_desc` funcs to the `FunctionsList` map on start when the engine is created in `promql/engine.go`?

roidelapluie commented:
I think a proper error saying that the feature needs to be enabled if the feature flag is not set would be better. However, I don't know if we can properly error from promql functions so we might need to be creative.

galexrt commented:
@roidelapluie I think I have found a way to check the feature flag in the evaluator. Please take another look at the latest code changes, thanks!

galexrt commented:
@roidelapluie I have updated the PR to resolve the conflicts. Any news on this?

juliusv commented on docs/feature_flags.md:
```diff
@@ -127,10 +127,18 @@ corresponding native histogram. However, if a native histogram is present,
 Prometheus will ignore the corresponding conventional histogram, with the
 notable exception of exemplars, which are always ingested.
 
+
```
Unrelated new empty line?

juliusv commented on promql/engine.go:
```diff
@@ -1386,6 +1394,8 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, annotations.Annotatio
 				return ev.rangeEvalTimestampFunctionOverVectorSelector(vs, call, e)
 			}
+		} else if (e.Func.Name == "sort_by_label" || e.Func.Name == "sort_by_label_desc") && !ev.enableSortByLabel {
+			ev.error(fmt.Errorf("sort_by_label and sort_by_label_desc feature flag not enabled"))
```
This makes it sound as if `sort_by_label` and `sort_by_label_desc` are the feature flag names, how about something like this:

```promql
sort_by_label() and sort_by_label_desc() require the "promql-sort-by-label" feature flag to be set
```

juliusv reviewed (commented):
Thank you, looks pretty good to me now!

It seems like the main outstanding question is around sort stability and whether we should resort to:

* Sorting by sample value
* Sorting by the remaining label set
* Do nothing

I would be for sorting by the remaining label set if that is easily doable (hopefully just a small extension of the SortFunc for the `0` string comparison case? Sorting by sample value seems too unrelated to the label sorting and is also not necessarily always stable, as two series can and often do have the same sample value.

juliusv commented on documentation/examples/prometheus.yml:
```diff
@@ -20,10 +20,10 @@ rule_files:
 
     static_configs:
-      - targets: ["localhost:9090"]
+      - targets: ["localhost:9138"]
```
Looks like some other unrelated changes crept into the PR here now?

galexrt replied:
I have removed this unrelated change, PTAL

juliusv commented:
Thanks! I see the tests failing with this error currently:

```
--- FAIL: TestEvaluations (5.87s)
    --- FAIL: TestEvaluations/testdata/functions.test (1.20s)
        test.go:105: 
            	Error Trace:	/__w/prometheus/prometheus/promql/test.go:105
            	            				/__w/prometheus/prometheus/promql/test.go:83
            	Error:      	Received unexpected error:
            	            	error in eval sort_by_label(http_requests, "instance", "group") (line 498): expected metric {__name__="http_requests", group="production", instance="0", job="api-server"} with [100.000000] at position 3 but was at 1
            	Test:       	TestEvaluations/testdata/functions.test
```

juliusv commented on documentation/examples/prometheus.yml:
```diff
@@ -20,7 +20,7 @@ rule_files:
 scrape_configs:
   # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
-  - job_name: "prometheus"
+  - job_name: "extended-ceph-exporter"
```
I still see this unrelated change in here?

galexrt replied:
Fixed, all unrelated changes should be gone now

galexrt commented:
> Thanks! I see the tests failing with this error currently:
> 
> ```
> --- FAIL: TestEvaluations (5.87s)
>     --- FAIL: TestEvaluations/testdata/functions.test (1.20s)
>         test.go:105: 
>             	Error Trace:	/__w/prometheus/prometheus/promql/test.go:105
>             	            				/__w/prometheus/prometheus/promql/test.go:83
>             	Error:      	Received unexpected error:
>             	            	error in eval sort_by_label(http_requests, "instance", "group") (line 498): expected metric {__name__="http_requests", group="production", instance="0", job="api-server"} with [100.000000] at position 3 but was at 1
>             	Test:       	TestEvaluations/testdata/functions.test
> ```

I have changed the sorting logic slightly, and it should now work better when more than 2 labels are involved.
To note sorting with 3 or more labels is "weird" because if the first label doesn't it doesn't compare the next label(s), added some test cases for that as well.

PTAL

roidelapluie commented:
Thanks!

pstibrany commented:
It's amazing to see this merged, 9 years after first https://github.com/prometheus/prometheus/pull/1533 :) Thanks @galexrt and all the reviewers! 🎉

bobrik commented:
Straightforward string sorting yield very questionable results for humans:

<img width="609" alt="image" src="https://github.com/prometheus/prometheus/assets/89186/e817418a-ab72-4193-8e90-467c079b24e3">

---

I urge you to consider using natural sorting instead.

Computers can sort by label however they want even without using `sort_by_label`, but humans need a little help.

See these PRs that added natural sorting to Grafana:

* https://github.com/grafana/grafana/pull/78024
* https://github.com/grafana/grafana/pull/78494

juliusv commented:
@bobrik Agreed, good point! It's not yet too late to change things, since the new sorting functions are added under an experimental feature flag. Do you want to send a PR?

bobrik commented:
Here's my attempt: https://github.com/prometheus/prometheus/pull/13411.
